### PR TITLE
Fix task leaks

### LINKED ISSUE #2378: A lot of tasks remain idle
**Issue summary**

Looking at tokio-console a lot of `tipset-state` tasks from tipset syncer stick to `IDLE` State, but this should not happen.

This results in the `tasks have lost their waker` warning.

idle_tasks

**Acceptance Criteria**

- [ ] We can run forest and the number of `IDLE` tasks should stay low (<150)
- [ ] Those tasks should end with the `DONE` State

### PR DESCRIPTION
**Pull request by elmattic**
**Summary of changes**
Changes introduced in this pull request:
- Add a new tipset state cache
- Use a Tokio `Mutex` for tipset cache and skip cache
- A panic in Tokio tasks does end Forest execution

**Reference issue to close (if applicable)**
<!-- Include the issue reference this pull request is connected to -->
<!--(e.g. Closes #1)-->
Closes #2378

This PR also helps #2278, before I had to stop the node and restart it, now Forest can cope with the errors.

This should also helps with #1891, but I would need to test it more.

**Other information and links**
<!-- Add any other context about the pull request here. -->

After a few hours I have now:

```
connection: http://127.0.0.1:6669/ (CONNECTED) dropped: 238899318 async_ops, 220376 tasks, 66281 resources
views: t = tasks, r = resources
controls: ←→ or h, l = select column (sort), ↑↓ or k, j = scroll, ↵ = view details, i = invert sort (highest/lowest), q = quit gg = scroll to top, G = scroll to bottom
╭Warnings──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│⚠ 2 tasks have lost their waker                                                                                                                                                                                                                       │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭Tasks (14101) ▶ Running (0) ⏸ Idle (42)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│Warn  ID    State  Name  Total      Busy▿      Idle       Polls     Target                Location                                                                            Fields                                                                  │
│⚠ 1   3873  ⏸            18221.0024 2173.3171s 16047.6853 388539626 tokio::task           forest/daemon/src/main.rs:205:37                                                    kind=block_on task.id=9                                                 │
│      10412 ⏸            18220.5926 1983.4843s 16237.1084 12668093  tokio::task           forest/daemon/src/daemon.rs:378:14                                                  kind=task task.id=65                                                    │
│       5281 ⏸            18220.9678  205.5302s 18015.4376 2011781   tokio::task           forest/daemon/src/daemon.rs:307:14                                                  kind=task task.id=16                                                    │
│       6940 ⏸            17915.9303   77.8546s 17838.0756 712206    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=12478                                                 │
│      10316 ⏸            18217.5924   72.0815s 18145.5109 673735    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=224                                                   │
│       7612 ⏸            18157.8372   70.3051s 18087.5321 675787    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=2972                                                  │
│       4649 ⏸            17747.0315   69.9903s 17677.0411 848617    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=15758                                                 │
│       6769 ⏸            18113.9681   68.2304s 18045.7377 632219    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=4855                                                  │
│        890 ⏸            18160.1103   67.6911s 18092.4192 589315    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=2707                                                  │
│       7294 ⏸            16351.6285   61.8610s 16289.7675 562499    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=37769                                                 │
│       8488 ⏸            18092.8404   61.3420s 18031.4985 646594    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=5949                                                  │
│       5083 ⏸            18219.9676   57.3940s 18162.5736 569865    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=91                                                    │
│        365 ⏸            18115.1672   57.1726s 18057.9946 586809    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=4832                                                  │
│       8326 ⏸            18032.8759   56.7663s 17976.1096 563911    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=8317                                                  │
│      10535 ⏸            18016.1365   55.9967s 17960.1399 546304    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=9116                                                  │
│      10369 ⏸            18219.0330   53.4231s 18165.6100 526715    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=125                                                   │
│       1057 ⏸            18093.7608   53.3261s 18040.4347 467343    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=5912                                                  │
│       3349 ⏸            18056.0239   53.1697s 18002.8542 489736    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=7452                                                  │
│       9146 ⏸            18035.0298   53.0919s 17981.9378 541475    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=8298                                                  │
│       6378 ⏸            18082.4247   52.7955s 18029.6291 492856    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=6299                                                  │
│       3031 ⏸            18206.5392   51.0970s 18155.4421 512492    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=571                                                   │
│       7291 ⏸            18148.2528   50.7075s 18097.5453 492429    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=3437                                                  │
│       4055 ⏸            16641.2308   49.7276s 16591.5032 466920    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=34543                                                 │
│      12343 ⏸            18048.0098   49.1320s 17998.8777 405557    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=7750                                                  │
│       9481 ⏸            18063.9426   47.0713s 18016.8713 442815    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=7277                                                  │
│       7318 ⏸            18013.3025   46.8351s 17966.4674 384884    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=9265                                                  │
│       9533 ⏸            18110.0727   46.5353s 18063.5374 430756    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=5003                                                  │
│       7069 ⏸            14237.0910   44.2155s 14192.8755 414034    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=75208                                                 │
│      11398 ⏸            17866.6740   44.0037s 17822.6703 414580    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=13889                                                 │
│      13362 ⏸            12900.8267   43.4709s 12857.3558 390342    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=97113                                                 │
│       9844 ⏹            14968.0694   43.4698s 14924.5996 436127    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=10509                                                 │
│       1639 ⏹            13935.6827   38.7954s 13896.8872 364099    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=33502                                                 │
│       2788 ⏸            11795.0214   36.6565s 11758.3649 354072    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=117402                                                │
│        926 ⏸            9618.1299s   32.9547s 9585.1752s 353460    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=153805                                                │
│        322 ⏸            10286.2164   31.5000s 10254.7164 296318    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=142058                                                │
│      13212 ⏸            11798.7928   31.1060s 11767.6868 306066    tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=117366                                                │
│         36 ⏹              12.2506s   12.2505s  57.0450µs 1         tokio::task::blocking <unknown location>                                                                  fn=forest_state_manager::StateManager<forest_db::parity_db::ParityDb>::c│
│       5883 ⏹              10.1610s   10.1144s  46.6344ms 24        tokio::task           /home/guillaume/Code/forest/blockchain/chain_sync/src/tipset_syncer.rs:1346:22      kind=task task.id=250537                                                │
│       1740 ⏹              10.1126s   10.1125s  45.9920µs 1         tokio::task::blocking <unknown location>                                                                  fn=forest_state_manager::StateManager<forest_db::parity_db::ParityDb>::c│
│       2884 ⏹               8.4962s    8.4962s  72.5180µs 1         tokio::task::blocking <unknown location>                                                                  fn=forest_state_manager::StateManager<forest_db::parity_db::ParityDb>::c│
│       4227 ⏸            2671.6492s    7.4592s 2664.1900s 77275     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=271972                                                │
│      11438 ⏹            1142.0618s    5.3341s 1136.7276s 47027     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=238817                                                │
│       8407 ⏸            1694.9840s    4.3468s 1690.6372s 45179     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=287714                                                │
│        344 ⏹             618.2931s    3.1109s  615.1823s 22702     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=263424                                                │
│       3151 ⏹            1101.9392s    2.8632s 1099.0760s 27242     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=232525                                                │
│       4722 ⏹             601.4936s    2.4475s  599.0461s 25620     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=275721                                                │
│        247 ⏹               2.2264s    2.2263s 139.1400µs 1         tokio::task::blocking <unknown location>                                                                  fn=forest_fil_cns::validation::validate_block<forest_db::parity_db::Pari│
│       9710 ⏹            1132.4639s    2.0975s 1130.3664s 22032     tokio::task           <cargo>/libp2p-swarm-0.41.1/src/executor.rs:42:17                                   kind=task task.id=283512                                                │
│       5850 ⏹               1.3746s    1.3739s 695.3560µs 1         tokio::task::blocking <unknown location>                                                                  fn=forest_fil_cns::validation::validate_block<forest_db::parity_db::Pari│
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

<!-- Thank you 🔥 -->

### CODE REVIEW

File: blockchain/state_manager/src/lib.rs [hunk 1]
```diff
@@ -313,64 +391,47 @@ where
...
+                        };
+
+                        // Write back to cache, release lock and return value
+                        self.cache.insert(key.clone(), cid_pair);
+                        Ok(cid_pair)
+                    }
+                }
+            }
         }
-        forest_metrics::metrics::LRU_CACHE_MISS
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:370]: So `LRU_CACHE_HIT` and `LRU_CACHE_MISS` are no longer tracked?

Comment (elmattic) [reply blockchain/state_manager/src/lib.rs:370]: My bad, will add them back.

**Comment by LesnyRumcajs** [comment]
Could you please explain why we need a `Mutex` instead of a `RwLock`?

**Comment by lemmih** [comment]
> Could you please explain why we need a `Mutex` instead of a `RwLock`?

Reading from an `LruCache` updates the access times and the cache must therefore be mutable. This makes the `read` part of a `read-write` lock unnecessary.

File: Cargo.toml [hunk 2]
```diff
@@ -219,3 +219,4 @@ lto = "off"
 [profile.release]
 # https://doc.rust-lang.org/cargo/reference/profiles.html#strip
 strip = true
+panic = "abort"
```
[the remark this answers is not in the document]
Comment (elmattic) [reply Cargo.toml:222]: Yes, this is a bit harsh. On the other hand we should have a way to catch panics. Right not as far as I understand, we catch panics only in cases we are testing for a `JoinError`:

https://docs.rs/tokio/1.23.0/tokio/task/struct.JoinError.html

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 3]
```diff
@@ -1114,18 +1114,19 @@ async fn sync_messages_check_state<
...
-            consensus.clone(),
-            state_manager.clone(),
-            &chainstore,
-            bad_block_cache,
-            full_tipset,
-            genesis,
-            invalid_block_strategy,
-        )
-        .await?;
-        timer.observe_duration();
```
[the remark this answers is not in the document]
Comment (elmattic) [reply blockchain/chain_sync/src/tipset_syncer.rs:1128]: Good catch, will check this.

File: blockchain/state_manager/src/lib.rs [hunk 4]
```diff
@@ -54,6 +56,84 @@ const DEFAULT_TIPSET_CACHE_SIZE: NonZeroUsize =
...
+    {
+        let mut lock = self.cache.lock().unwrap();
+        func(&mut lock)
+    }
+
+    pub fn get_status(&self, key: &TipsetKeys) -> Status {
+        self.with_inner(|inner| match inner.values.get(key) {
+            Some(v) => Status::Done(*v),
+            None => {
+                for (v, l) in inner.pending.iter() {
```
[the remark this answers is not in the document]
Comment (elmattic) [reply blockchain/state_manager/src/lib.rs:103]: Good idea, thanks.

File: blockchain/state_manager/src/lib.rs [hunk 5]
```diff
@@ -54,6 +56,84 @@ const DEFAULT_TIPSET_CACHE_SIZE: NonZeroUsize =
...
+    }
+
+    pub fn get(&self, key: &TipsetKeys) -> Option<CidPair> {
+        self.with_inner(|inner| inner.values.get(key).copied())
+    }
+
+    pub fn insert(&self, key: TipsetKeys, value: CidPair) {
+        self.with_inner(|inner| {
+            inner.values.put(key.clone(), value);
+            inner.pending.retain(|(k, _)| k != &key);
```
[the remark this answers is not in the document]
Comment (elmattic) [reply blockchain/state_manager/src/lib.rs:132]: Indeed. It's safe to do as long as `values` and `pending` are mutated inside the lock.

File: blockchain/state_manager/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (elmattic) [reply blockchain/state_manager/src/lib.rs]: I had already a `find` actually, so I just removed the useless for loop.

File: blockchain/chain_sync/src/tipset_syncer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (elmattic) [reply blockchain/chain_sync/src/tipset_syncer.rs:1128]: Yes, this is not needed because drop of the `_timer` will call `observe_duration` anyway.

**Review by LesnyRumcajs — APPROVED** [review]
rock-solid

File: blockchain/state_manager/src/lib.rs [hunk 6]
```diff
@@ -54,6 +56,99 @@ const DEFAULT_TIPSET_CACHE_SIZE: NonZeroUsize =
...
+    where
+        F: FnOnce(&mut TipsetStateCacheInner) -> T,
+    {
+        let mut lock = self.cache.lock().unwrap();
+        func(&mut lock)
+    }
+
+    pub async fn get_or_else<F, Fut>(&self, key: &TipsetKeys, cb: F) -> anyhow::Result<CidPair>
+    where
+        F: Fn(TipsetKeys) -> Fut,
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:101]: We can skip the key argument here, can't we?

```suggestion
        F: Fn() -> Fut,
```

File: blockchain/state_manager/src/lib.rs [hunk 7]
```diff
@@ -313,64 +406,47 @@ where
...
+
+        // if cache_miss {
+        //     forest_metrics::metrics::LRU_CACHE_MISS
+        //         .with_label_values(&[forest_metrics::metrics::values::STATE_MANAGER_TIPSET])
+        //         .inc();
+        // } else {
+        //     forest_metrics::metrics::LRU_CACHE_HIT
+        //         .with_label_values(&[forest_metrics::metrics::values::STATE_MANAGER_TIPSET])
+        //         .inc();
+        // }
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:448]: It's easier if we cache misses and the total number of queries.

File: blockchain/state_manager/src/lib.rs [hunk 8]
```diff
@@ -313,64 +406,47 @@ where
...
+                    // generic constants are not implemented yet this is a lowcost method for now
+                    let no_func =
+                        None::<fn(&Cid, &ChainMessage, &ApplyRet) -> Result<(), anyhow::Error>>;
+                    let ts_state = self.compute_tipset_state(tipset, no_func).await?;
+                    debug!("Completed tipset state calculation {:?}", tipset.cids());
+                    ts_state
+                };
+
+                // Write back to cache, release lock and return value
+                self.cache.insert(key.clone(), cid_pair);
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:435]: Shouldn't this be done in `get_or_else`?

Comment (elmattic) [reply blockchain/state_manager/src/lib.rs:435]: Indeed, I was hesitating to move it here, will do.

File: blockchain/state_manager/src/lib.rs [hunk 9]
```diff
@@ -313,64 +427,34 @@ where
...
-            (*tipset.parent_state(), *message_receipts.message_receipts())
-        } else {
-            // generic constants are not implemented yet this is a lowcost method for now
-            let no_func = None::<fn(&Cid, &ChainMessage, &ApplyRet) -> Result<(), anyhow::Error>>;
-            let ts_state = self.compute_tipset_state(tipset, no_func).await?;
-            debug!("completed tipset state calculation {:?}", tipset.cids());
-            ts_state
-        };
+        let key = tipset.key();
+        let result = self
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:431]: We can remove this let binding and immediately return the result, I guess.

**Review by lemmih — APPROVED** [review]

**Pull request merged**